### ISSUE TITLE
VideoTripletEngine multiple inheritance not working & possible fix

### ISSUE THREAD

**Issue by lennartmoritz** [KaiyangZhou/deep-person-reid#360]
### Error:
When I try to use triplet loss with a video dataset I get the following error from [Line 74 in engine/image/triplet.py](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/engine/image/triplet.py#L74):
```
Exception has occurred: TypeError
__init__() missing 1 required positional argument: 'optimizer'
  File "E:\GitHub\deep-person-reid\torchreid\engine\image\triplet.py", line 74, in __init__
    super(ImageTripletEngine, self).__init__(datamanager, use_gpu)
  File "E:\GitHub\deep-person-reid\torchreid\engine\video\triplet.py", line 89, in __init__
    label_smooth=label_smooth
  File "E:\GitHub\person-reid-in-drone-images\scripts\main.py", line 74, in build_engine
    label_smooth=cfg.loss.softmax.label_smooth
  File "E:\GitHub\person-reid-in-drone-images\scripts\main.py", line 179, in main
    engine = build_engine(cfg, datamanager, model, optimizer, scheduler)
  File "E:\GitHub\person-reid-in-drone-images\scripts\main.py", line 189, in <module>
    main()
```
_Just ignore that the paths throughout this error change from deep-person-reid to person-reid-in-drone-images. Its just an ugly setup structure of my project that I've been to busy to fix yet_
### Context
I belive multiple inheritance was implemented incorrectly as the class VideoTripletEngine inheits from ImageTripletEngine and VideoSoftmaxEngine in [line 7 in engine/video/triplet.py](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/engine/video/triplet.py#L7):
`class VideoTripletEngine(ImageTripletEngine, VideoSoftmaxEngine):`

### Fix
I belive this can be fixed in two steps. 
The first step is changing the \_\_init__() method from the **ImageTripletEngine** class at [Line 74 in engine/image/triplet.py](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/engine/image/triplet.py#L74) from
`super(ImageTripletEngine, self).__init__(datamanager, use_gpu)`
to
`super(ImageTripletEngine, self).__init__(datamanager, use_gpu, optimizer=optimizer)`

The second step is to adapt the signature the **ImageTripletEngine's** base class (**Engine**) to accomodate the signature change made in the first step. This is done in [Line 31 in engine/engine.py](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/engine/engine.py#L31) by changing
`def __init__(self, datamanager, use_gpu=True):`
to
`def __init__(self, datamanager, use_gpu=True, **kwargs):`

I haven't tested if my changes are going to break anything else, and I am yet to see if my fix will result in good training results as I've been having trouble with triplet loss even with an image dataset see #357. I'm testing this loss in a video dataset now and if I feel like my fix didn't work or cause some new problems I will update this issue when my training has finished.

### Edit
After running the following settings I had poor results. My training crashed after 66/110 Epochs but I did an evaluation every 10 epochs. Note that i was able to get a mAP of 38,3% by using osnet with softmax before.
As you can see the mAP quickly drops to about 1% after the first 10 fixbase epochs. It is interesting to see that the Train/acc also quickly dropped from ~70% to ~0% immediatly after the 10 fixbase epochs were finished, as well as the Train/loss_t dropping from ~6.5 to between 0.4-0.3 immedeatly while Train/loss_x rebounds from  ~2 to ~5.5. 
I can't explain this behaviour yet but it is possible that my fix was not able to correctly preserve the backpropagation and weight adjustment needed for combined loss. To me it seems like cross entropy loss was no longer taken into account for the weight adjustments after the first 10 fixbase epochs. Maybe i can try to lower the weight_t to avoid this behavior in the future.
![Anmerkung 2020-08-03 122911](https://user-images.githubusercontent.com/51822945/89173833-24069800-d585-11ea-9648-d6ea3108f41a.png)
![Anmerkung 2020-08-03 122851](https://user-images.githubusercontent.com/51822945/89173958-5fa16200-d585-11ea-9e47-47591f2c681e.png)

```
model:
  name: 'osnet_x1_0'
  pretrained: True
  resume: ''

data:
  type: 'video'
  root: '/deeplearning/datasets/reid'
  sources: ['pdestre']
  targets: ['pdestre']
  height: 256
  width: 128
  combineall: False
  transforms: ['random_flip', 'random_crop']
  save_dir: '/log/osnet_x1_0_pdestre_combined_cosinelr4'
  split_id: 0

loss:
  name: 'triplet'
  softmax:
    label_smooth: True
  triplet:
    margin: 0.3
    weight_t: 0.5
    weight_x: 1.

train:
  optim: 'amsgrad'
  lr: 0.0015
  max_epoch: 110
  batch_size: 8
  fixbase_epoch: 10
  open_layers: ['classifier']
  lr_scheduler: 'cosine'
  stepsize: [60, 100]

test:
  batch_size: 45
  dist_metric: 'euclidean'
  normalize_feature: True
  evaluate: False
  eval_freq: 10
  rerank: False
  visrank: False

sampler:
  train_sampler: 'RandomIdentitySampler'
  train_sampler_t: 'RandomIdentitySampler'
  num_instances: 4

video:
  seq_len: 14
  sample_method: 'random'
  pooling_method: 'avg'
```

**Comment by KaiyangZhou** [comment]
good spot

fixed now
